pkg/metrics/metric/metric.go

```diff
@@ -196,3 +239,85 @@ func (b Opts) GetConfigName() string {
 	}
 	return b.ConfigName
 }
+
+type Label struct {
+	Name string
+	// If defined, only these values are allowed.
+	Values Values
+}
+
+type Values map[string]struct{}
+
+func NewValues(vs ...string) Values {
+	vals := Values{}
+	for _, v := range vs {
+		vals[v] = struct{}{}
+	}
+	return vals
+}
+
+type Labels []Label
+
+func (lbls Labels) labelNames() []string {
+	lns := make([]string, len(lbls))
+	for i, label := range lbls {
+		lns[i] = label.Name
+	}
+	return lns
+}
+
+type labelSet struct {
+	lbls Labels
+	m    map[string]map[string]struct{}
+}
+
+func (l *labelSet) namesToValues() map[string]map[string]struct{} {
+	if l.m != nil {
+		return l.m
+	}
+	l.m = make(map[string]map[string]struct{})
+	for _, label := range l.lbls {
+		l.m[label.Name] = label.Values
+	}
+	return l.m
+}
+
+func (l labelSet) checkLabels(labels prometheus.Labels) error {
+	for name, value := range labels {
+		if lvs, ok := l.namesToValues()[name]; ok {
+			if _, ok := lvs[value]; !ok {
+				return fmt.Errorf("value %s not allowed for label %s (should be one of %v)", value, name, lvs)
+			}
+		} else {
+			return fmt.Errorf("invalid label name: %s", name)
+		}
+	}
+	return nil
+}
+
+func (l labelSet) checkLabelValues(lvs []string) error {
+	if len(l.lbls) != len(lvs) {
+		return fmt.Errorf("invalid labels")
+	}
+	for i, label := range l.lbls {
+		if label.Values != nil {
```

Comment:
nit: Don't think we need to check for nil map here?

Reply:
wanted to avoid crashing in case someone doesn't set this value.

Reply:
I think you can index into a nil map